start_admin_sshd.sh

```diff
@@ -6,38 +6,94 @@ set -e
-    continue
-  fi
-  echo "${public_key_data}" >> "${ssh_authorized_keys}"
-  if ! grep -q "${public_key_data}" "${ssh_authorized_keys}"; then
-    echo "Failed to write key with index ${public_key_index} to authorized_keys" >&2
-    continue
+declare -r ssh_host_key_dir="/.bottlerocket/host-containers/admin/etc/ssh"
+declare -r ssh_config_dir="/home/ec2-user/.ssh"
+declare -r user_data="/.bottlerocket/host-containers/admin/user-data"
+available_auth_methods=0
+
+parse_authorized_keys() {
+    # Extract the keys from user-data json
+    local raw_keys
+    if ! raw_keys=$(jq -e -r '.["ssh"]["authorized_keys"][]' "${user_data}" 2>/dev/null); then
+      echo "Failed to parse authorized_keys from ${user_data}" >&2
+      return 1
+    fi
+
+    # Map the keys to avoid improper splitting
+    local mapped_keys
+    mapfile -t mapped_keys <<< "${raw_keys}"
+
+    # Verify the keys are valid
+    local key
+    local -a valid_keys
+    for key in "${mapped_keys[@]}"; do
+      if ! echo "${key}" | ssh-keygen -lf - &>/dev/null; then
+        echo "Failed to validate ${key}" >&2
+        continue
+      fi
+      valid_keys+=( "${key}" )
+    done
+
+    ( IFS=$'\n'; echo "${valid_keys[*]}" )
+}
+
+# Populate authorized_keys with all the authorized keys found in user-data
+if authorized_keys=$(parse_authorized_keys); then
+  ssh_authorized_keys="${ssh_config_dir}/authorized_keys"
+  install -D -m 0600 /dev/null "${ssh_authorized_keys}"
+  echo "${authorized_keys}" > "${ssh_authorized_keys}"
+  ((++available_auth_methods))
+fi
+
+parse_trusted_user_ca_keys() {
+    # Extract the keys from user-data json
+    local raw_keys
+    if ! raw_keys=$(jq -e -r '.["ssh"]["trusted_user_ca_keys"][]' "${user_data}" 2>/dev/null); then
+      echo "Failed to parse trusted_user_ca_keys from ${user_data}" >&2
+      return 1
+    fi
+
+    # Map the keys to avoid improper splitting
+    local mapped_keys
+    mapfile -t mapped_keys <<< "${raw_keys}"
+
+    # Verify the keys are valid
+    local key
+    local -a valid_keys
+    for key in "${mapped_keys[@]}"; do
+      if ! echo "${key}" | ssh-keygen -lf - &>/dev/null; then
+        echo "Failed to validate ${key}" >&2
+        continue
+      fi
+      valid_keys+=( "${key}" )
+    done
+
+    ( IFS=$'\n'; echo "${valid_keys[*]}" )
+}
+
+# Populate trusted_user_ca_keys with all the authorized keys found in user-data
+if trusted_user_ca_keys=$(parse_trusted_user_ca_keys); then
+  ssh_trusted_user_ca_keys="/etc/ssh/trusted_user_ca_keys.pub"
+  sshd_config="/etc/ssh/sshd_config"
+  install -D -m 0600 /dev/null "${ssh_trusted_user_ca_keys}"
+  echo "${trusted_user_ca_keys}" > "${ssh_trusted_user_ca_keys}"
+
+  cat <<EOF >> "${sshd_config}"
+
+# Configured by user data
+TrustedUserCAKeys ${ssh_trusted_user_ca_keys}
+EOF
+
+  if [[ -f "${ssh_authorized_keys}" ]]; then
+    echo "AuthorizedKeysFile ${ssh_authorized_keys}" >> "${sshd_config}"
   fi
```

Comment:
I don't think we should be configuring an absolute path to the authorized keys. I haven't tried this myself, but I'd suspect that an absolute path would authorize *all users* with the given file, meaning any user could login as any user. Even so, `sshd_config(5)` has this to say:


<pre>
AuthorizedKeysFile

  Specifies the file that contains the public keys used for user authentication.
  The format is described in the AUTHORIZED_KEYS FILE FORMAT section of sshd(8).
  Arguments to AuthorizedKeysFile accept the tokens described in the TOKENS
  section. After expansion, AuthorizedKeysFile is taken to be an absolute path
  or one relative to the user's home direc‐ tory. Multiple files may be listed,
  separated by whitespace. Alternately this option may be set to none to skip
  checking for user keys in files. 

  The default is ".ssh/authorized_keys .ssh/authorized_keys2".
</pre>

I'm of a mind that we do not try to set or modify the option. The location we're installing the keys will work without adding the explicit path.